### PR TITLE
Terraform: Dev

### PR DESCRIPTION
**Pull request by danieltharp**
This is a working terraform config for DEV ONLY and documentation on how to deploy it.

I highly suggest we squash merge this bad boy as the intervening steps to working aren't really relevant.

### CODE REVIEW

File: docs/deployment.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# Remote Deployment
+
+This document will cover the process of deploying Wikijump on Amazon Web Services.
+
+## General Architecture
+
+Communication with Wikijump begins at DNS. If the address is for a wikijump.com subdomain, or a custom domain CNAMEd to wikijump.com, it will hit an Elastic Load Balancer. Specifically, it will resolve to the static (elastic) IP address of a Network Load Balancer listening on ports 80 and 443, but not terminating any SSL connections. It will proxy all this traffik to the Traefik edge proxy containers running in Elastic Container Service. Traefik handles the business of SSL termination and the permanent redirecting of http to https. It will then act as a reverse proxy, forwarding the request on to other containers running php-fpm and nginx. These containers are connected to a caching layer (memcached) and a database (postgres).
```
Comment (emmiegit) [inline docs/deployment.md]: Not blocking: would suggest wrapping words like `wikijump.com` or `CNAME` in backticks for readability. also `http` -> `HTTP`

Comment (emmiegit) [inline docs/deployment.md]: How can an IP address be both static and elastic?

Comment (emmiegit) [inline docs/deployment.md]: > it will hit an Elastic Load Balancer. Specifically, it will resolve to the static (elastic) IP address of a Network Load Balancer listening on ports 80 and 443

This seems unclear. Is there an Elastic Load Balancer in front of the Network Load Balancer, or does it have a "load balancer" (generic), whose specific subtype is "Network Load Balancer"?

File: docs/deployment.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+If the address is for a wjfiles.com subdomain, it will instead hit a CloudFront distribution. This will terminate SSL and examine the requested path. If it is for a file asset, it will retrieve the file from S3. If it is for a code-type asset, it will proxy the request to Traefik, which goes to the same php-fpm and nginx containers to return the code. As objects retrieved in this way should generally be static assets, we will make use of caching to reduce the load on our internals.
```
Comment (emmiegit) [inline docs/deployment.md]: Do `wjfiles.com` subdomains permit HTTP?

Edit: looking below it seems it's HTTPS-only

File: docs/deployment.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Prerequisites
+
+Deployment was designed to need a minimal amount of work done in advance, but there is always some. Feel free to contribute code to run `aws` CLI calls for some of this.
```
Comment (emmiegit) [inline docs/deployment.md]: Would reword a bit, something like:
```suggestion
Deployment was designed to need a minimal amount of work done in advance, but there is always some. Feel free to contribute code to run automate this setup (e.g. a script that calls the `aws` CLI).
```

**Comment by danieltharp** [comment]
I've updated the PR to also make use of a new GitHub Actions file, if this one works I'll build a matching one for postgres.

File: infra/terraform/dev/deploy/.gitignore [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+*.tfstate.*
+
+# Crash log files
+crash.log
+
+# Ignore any .tfvars files that are generated automatically for each Terraform run. Most
+# .tfvars files are managed as part of configuration and so should be included in
+# version control.
+#
+# example.tfvars
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/.gitignore:15]: This is commented out. Intentional?

File: infra/terraform/dev/deploy/.gitignore [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+override.tf.json
+*_override.tf
+*_override.tf.json
+
+# Include override files you do wish to add to version control using negated pattern
+#
+# !example_override.tf
+
+# Include tfplan files to ignore the plan output of command: terraform plan -out=tfplan
+# example: *tfplan*
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/.gitignore:29]: Ditto

File: infra/terraform/dev/deploy/acm.tf [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  ]
+  validation_method = "DNS"
+
+  lifecycle {
+    create_before_destroy = true
+  }
+
+  # We hardcode us-east-1 here because CloudFront will look for this cert in us-east-1.
+  # See: https://docs.aws.amazon.com/acm/latest/userguide/acm-regions.html
+  provider = aws.us-east-1
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/acm.tf:14]: pressing f to pay respects

File: infra/terraform/dev/deploy/cloudfront.tf [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      origin_access_identity = aws_cloudfront_origin_access_identity.s3_oai.cloudfront_access_identity_path
+    }
+    # custom_origin_config {
+    #   http_port                = 80
+    #   https_port               = 443
+    #   origin_protocol_policy   = "http-only"
+    #   origin_ssl_protocols     = ["TLSv1.2"]
+    #   origin_keepalive_timeout = 15
+    #   origin_read_timeout      = 30
+    # }
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/cloudfront.tf]: Commented out code.

File: infra/terraform/dev/deploy/cloudfront.tf [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    path_pattern     = "local--code/*"
+    allowed_methods  = ["GET", "HEAD", "OPTIONS"]
+    cached_methods   = ["GET", "HEAD"]
+    target_origin_id = aws_lb.wikijump_elb.dns_name
+
+    forwarded_values {
+      query_string = false
+
+      cookies {
+        forward = "all"
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/cloudfront.tf:78]: Why does code send cookies and files don't?

File: infra/terraform/dev/deploy/cloudwatch.tf [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  retention_in_days = "7"
+}
+resource "aws_cloudwatch_log_group" "php-fpm" {
+  name              = "ecs/php-fpm-${var.environment}"
+  retention_in_days = "7"
+}
+resource "aws_cloudwatch_log_group" "traefik" {
+  name              = "ecs/traefik-${var.environment}"
+  retention_in_days = "7"
+}
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/cloudwatch.tf:19]: Not blocking: no lines between resource blocks

File: infra/terraform/dev/deploy/ecs.tf [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  name_prefix           = "CLUSTER_NODES_"
+  desired_capacity      = 1
+  max_size              = 1
+  min_size              = 1
+  vpc_zone_identifier   = [aws_subnet.container_subnet.id]
+  protect_from_scale_in = false
+
+  mixed_instances_policy {
+    instances_distribution {
+      on_demand_percentage_above_base_capacity = 0
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/ecs.tf:21]: Note: as someone not very familiar with AWS, this is the part where I'm reading through where I start losing understanding of what this means.

If there is any relevant documentation about what these settings mean that could be linked in a comment, that could help with readability.

File: infra/terraform/dev/deploy/ecs.tf [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+      minimum_scaling_step_size = 1
+      status                    = "ENABLED"
+      target_capacity           = 1
+    }
+  }
+}
+
+resource "aws_ecs_task_definition" "wikijump_task" {
+  family = "wikijump-${var.environment}-ec2"
+  # container_definitions    = file("task-definitions/dev-ec2.json")
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/ecs.tf]: Commented code

File: infra/terraform/dev/deploy/ecs.tf [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  container_definitions    = "[${module.cache.json_map_encoded},${module.database.json_map_encoded},${module.php-fpm.json_map_encoded},${module.reverse-proxy.json_map_encoded}]"
+  requires_compatibilities = ["EC2"]
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/ecs.tf:62]: Note: This part I'm starting to understand better again

File: infra/terraform/dev/deploy/ecs.tf [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  }
+  load_balancer {
+    target_group_arn = aws_lb_target_group.elb_target_group_80.arn
+    container_name   = "reverse-proxy"
+    container_port   = 80
+  }
+}
+
+
+
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/ecs.tf]: Not blocking: 2+ lines between resources, extraneous newlines at the end.

File: infra/terraform/dev/deploy/ecs.tf [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    }
+  }
+}
+
+
+resource "aws_ecs_service" "wikijump" {
+  name                 = "wikijump-${var.environment}-svc"
+  cluster              = aws_ecs_cluster.wikijump-ecs.id
+  task_definition      = aws_ecs_task_definition.wikijump_task.arn
+  desired_count        = 1 # This will be a var as we grow
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/ecs.tf:88]: :+1:

File: infra/terraform/dev/deploy/elb.tf [hunk 15]
```diff
@@ -0,0 +1,66 @@
+# ELB
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/elb.tf:1]:
```suggestion
# Elastic Load Balancer
```

File: infra/terraform/dev/deploy/elb.tf [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+resource "aws_lb" "wikijump_elb" {
+  name               = "wikijump-public-elb-${var.environment}"
+  internal           = false
+  load_balancer_type = "network"
+  subnet_mapping {
+    subnet_id     = aws_subnet.elb_subnet.id
+    allocation_id = aws_eip.elb_eip.id
+  }
+  # ip_address_type = "dualstack"
+  # Enable this once stable.
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/elb.tf]:
```suggestion
  # TODO: Enable this once stable.
```

File: infra/terraform/dev/deploy/elb.tf (hunk 15 above)
Comment (emmiegit) [reply infra/terraform/dev/deploy/elb.tf:1]: Non-blocking obviously, but worth considering if you want any kind of consistent comment header at the top of terraform files going forward.

File: infra/terraform/dev/deploy/iam.tf [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+# Execution IAM Role
+#####
+resource "aws_iam_role" "execution" {
+  name               = "wj-${var.environment}-execution-role"
+  assume_role_policy = data.aws_iam_policy_document.assume_role_policy.json
+}
+
+resource "aws_iam_role_policy_attachment" "ecs_task_execution_role_policy_attach" {
+  role       = aws_iam_role.execution.name
+  policy_arn = "arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/iam.tf:11]: Why does this use a hardcoded string and this next resource uses a variable?

File: infra/terraform/dev/deploy/launch.tf [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  dynamic "part" {
+    for_each = var.user_data
+    content {
+      content_type = "text/x-shellscript"
+      content      = part.value
+    }
+  }
+}
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/launch.tf:23]: I don't really understand what this does. It sends that bash script to some new server?

File: infra/terraform/dev/deploy/launch.tf [hunk 19]
```diff
@@ -0,0 +1,35 @@
+data "template_cloudinit_config" "config" {
+  gzip          = false
+  base64_encode = true
+
+  part {
+    content_type = "text/x-shellscript"
+    content      = <<EOT
+#!/bin/bash
+echo ECS_CLUSTER="wikijump-${var.environment}" >> /etc/ecs/ecs.config
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/launch.tf]: Since this is a terraform variable and not bash, it should use single quotes:
```suggestion
echo ECS_CLUSTER='wikijump-${var.environment}' >> /etc/ecs/ecs.config
```

File: infra/terraform/dev/deploy/launch.tf [hunk 20]
```diff
@@ -0,0 +1,35 @@
+data "template_cloudinit_config" "config" {
+  gzip          = false
+  base64_encode = true
+
+  part {
+    content_type = "text/x-shellscript"
+    content      = <<EOT
+#!/bin/bash
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/launch.tf:8]: Always recommend setting these on scripts to make soft-warnings into errors.

* `set -e` -- fail if any command does
* `set -u` -- fail if a bash variable is unset when used

```suggestion
#!/bin/bash
set -eu
```

File: infra/terraform/dev/deploy/providers.tf [hunk 21]
```diff
@@ -0,0 +1,18 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 3.0"
+    }
+  }
+}
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/providers.tf:8]: Not blocking:
```suggestion
}

```

File: infra/terraform/dev/deploy/route53.tf [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+    zone_id                = aws_lb.wikijump_elb.zone_id
+    evaluate_target_health = true
+  }
+
+  allow_overwrite = true
+}
+
+resource "aws_route53_record" "primary_wildcard" {
+  zone_id         = aws_route53_zone.primary.zone_id
+  name            = "*.${var.web_domain}"
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/route53.tf:24]: You can `CNAME` wildcards? Huh.

File: infra/terraform/dev/deploy/s3.tf [hunk 23]
```diff
@@ -0,0 +1,9 @@
+resource "aws_s3_bucket" "elb_logs" {
+  bucket = "wikijump-elb-logs-${var.environment}"
+  acl    = "log-delivery-write"
+}
+
+resource "aws_s3_bucket" "wikijump_assets" {
+  bucket = "wikijump-assets-${var.environment}"
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/s3.tf:7]: What are "assets" here? Code and files?

File: infra/terraform/dev/deploy/task-definitions.tf [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+      "awslogs-stream-prefix" = "ecs"
+    }
+  }
+}
+
+module "php-fpm" {
+  source = "github.com/cloudposse/terraform-aws-ecs-container-definition?ref=0.46.0"
+
+  container_name   = "php-fpm"
+  container_image  = "${data.aws_ssm_parameter.WEB_ECR_URL.value}:develop"
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/task-definitions.tf:43]: Not blocking: May be worth making a variable for the repo branch name.

File: infra/terraform/dev/deploy/task-definitions.tf [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+
+  docker_labels = {
+    "traefik.enable"                                = "true"
+    "traefik.http.routers.php-fpm.rule"             = "Host(`${var.web_domain}`,`www.${var.web_domain}`,`${var.files_domain}`,`www.${var.files_domain}`)"
+    "traefik.http.routers.php-fpm.tls"              = "true"
+    "traefik.http.routers.php-fpm.tls.certresolver" = "mytlschallenge"
+  }
+
+  healthcheck = {
+    command     = ["CMD-SHELL", "curl -f http://localhost || exit 1"]
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/task-definitions.tf]: `|| exit 1` is redundant, it's already outputting non-zero

File: infra/terraform/dev/deploy/vpc.tf [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+  assign_generated_ipv6_cidr_block = true
+}
+
+# Subnets
+
+resource "aws_subnet" "elb_subnet" {
+  vpc_id                  = aws_vpc.wikijump_vpc.id
+  cidr_block              = var.elb_subnet
+  map_public_ip_on_launch = true
+  availability_zone       = format("%s%s", var.region, var.availability_zone)
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/vpc.tf:16]: Curious, is there a difference between this and `"${var.region}${var.availability_zone}"`?

File: infra/terraform/dev/deploy/vpc.tf [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+resource "aws_security_group_rule" "egress" {
+  from_port         = 0
+  to_port           = 0
+  protocol          = "-1"
+  cidr_blocks       = ["0.0.0.0/0"]
+  security_group_id = aws_security_group.ecs_nodes.id
+  type              = "egress"
+}
+
+resource "aws_security_group_rule" "ingress_self" {
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/vpc.tf:105]: Curious, why is it `egress` but then `ingress_self`?

File: infra/terraform/dev/deploy/wikijump.auto.tfvars [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+
+# VPC
+vpc_cidr_block   = "10.106.0.0/16"
+elb_subnet       = "10.106.0.0/24"
+container_subnet = "10.106.10.0/24"
+
+# Cloudfront/ELB
+# The cf_auth_token var isn't *sensitive* as such, but an attacker could add this header value to bypass Cloudfront and hit our load balancer directly.
+# So, you may as well store this as a secret, perhaps in Terraform Cloud's vars as a sensitive value so it doesn't show up in tf plans.
+# It can be anything, something vaguely random will work fine like a GUID.
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/wikijump.auto.tfvars]: Wording:
```suggestion
# It can be anything, something vaguely random like a GUID will work fine.
```

File: infra/terraform/dev/deploy/wikijump.auto.tfvars [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+
+ecs_php_memory = 768
+ecs_php_cpu    = 1024
+
+ecs_traefik_memory = 512
+ecs_traefik_cpu    = 512
+ecs_traefik_image  = "traefik:v2.3"
+
+# Misc
+letsencrypt_email = "info@wikijump.com"
```
Comment (emmiegit) [inline infra/terraform/dev/deploy/wikijump.auto.tfvars:36]: Minor, would suggest something like `security@wikijump.com`.

File: infra/terraform/dev/init/.gitignore [hunk 30]
```diff
@@ -0,0 +1,32 @@
+# Local .terraform directories
```
Comment (emmiegit) [inline infra/terraform/dev/init/.gitignore:1]: This looks similar to the other terraform gitignore file. Any chance / is it reasonable to merge them, and have these files reference the same with symlinks? Or just move it to a higher directory.

File: infra/terraform/dev/init/init.tf [hunk 31]
```diff
@@ -0,0 +1,50 @@
...
+      source  = "hashicorp/aws"
+      version = "~> 3.0"
+    }
+  }
+}
+# Configure the AWS Provider
+provider "aws" {
+  region = local.region
+}
+
```
Comment (emmiegit) [inline infra/terraform/dev/init/init.tf:18]: Not blocking, extra newline:
```suggestion
```

File: infra/terraform/prod/cloudfront.tf [hunk 32]
```diff
@@ -1,46 +1,96 @@
...
-    compress                = true
-    max_ttl                 = 60
-    }
-
-    viewer_certificate {
-        acm_certificate_arn = aws_acm_certificate.cf_wildcard_cert.arn
-  }
-}
\ No newline at end of file
+# resource "aws_cloudfront_distribution" "wikijump_cf_distro" {
```
Comment (emmiegit) [inline infra/terraform/prod/cloudfront.tf:1]: Commented code. 

If we're just disabling this file, maybe rename it? `cloudfront.tf -> cloudfront.disabled_tf`

File: infra/terraform/prod/ecs.tf [hunk 33]
```diff
@@ -1,26 +1,27 @@
-resource "aws_ecs_service" "wikijump" {
-    name    = "wikijump-${var.environment}"
-    cluster = aws_ecs_cluster.wikijump_ecs_cluster.id
-    task_definition = aws_ecs_task_definition.wikijump_task.arn
-    desired_count   = 1  # This will be a var as we grow
+# resource "aws_ecs_service" "wikijump" {
```
Comment (emmiegit) [inline infra/terraform/prod/ecs.tf:1]: Ditto

File: infra/terraform/prod/elb.tf [hunk 34]
```diff
@@ -3,13 +3,12 @@
...
-    load_balancer_type          = "application"
-    security_groups             = [aws_security_group.elb_sg.id]
+    load_balancer_type          = "network"
     subnet_mapping {
         subnet_id               = aws_subnet.elb_subnet.id
         allocation_id           = aws_eip.elb_eip.id
     }
-    
+    # ip_address_type = "dualstack"
     # Enable this once stable.
```
Comment (emmiegit) [inline infra/terraform/prod/elb.tf:12]:
```suggestion
    # TODO: Enable this once stable.
```

File: infra/terraform/prod/elb.tf [hunk 35]
```diff
@@ -21,86 +20,96 @@ resource "aws_lb" "wikijump_elb" {
...
-        values              = [var.cf_auth_token]
-        }
+        target_group_arn = aws_lb_target_group.elb_target_group_443.arn
     }
 }
 
-resource "aws_lb_listener_rule" "fallback" {
-    listener_arn            = aws_lb_listener.elb_listener.arn
-    priority                = 999
+# resource "aws_lb_listener_rule" "cloudfront_forward_80" {
```
Comment (emmiegit) [inline infra/terraform/prod/elb.tf:65]: Commented code.

Comment (emmiegit) [reply infra/terraform/prod/elb.tf:65]: For the record, I'm fine with large commented blocks of code if there is documentation explaining *why* it's commented, e.g. temporarily disabled, broken and needs a certain kind of fix, going to be deprecated, etc.

Commented out code in general makes me concerned it was a temporary development change that ends up in master, without clarity as to why it was commented out.

File: infra/terraform/prod/route53.tf [hunk 36]
```diff
@@ -48,3 +48,26 @@ resource "aws_route53_record" "files_wildcard" {
   records = [var.files_domain]
   allow_overwrite = true
 }
+
```
Comment (emmiegit) [inline infra/terraform/prod/route53.tf:51]: Not blocking, extra line:
```suggestion
```

File: infra/terraform/prod/vpc.tf [hunk 37]
```diff
@@ -39,11 +46,11 @@ resource "aws_internet_gateway" "wikijump_igw" {
 
 # Routes
 
-# # TODO: Add as needed and as will improve security posture.
+
```
Comment (emmiegit) [inline infra/terraform/prod/vpc.tf:49]: Why was this comment removed? Should there still be a `Routes` section?

File: install/docker-compose.aws-dev.yaml [hunk 38]
```diff
@@ -0,0 +1,60 @@
...
+  cache:
+    build:
+      context: ./cache
+    image: wikijump/memcached
+    restart: always
+  database:
+    build:
+      context: ./postgres
+    restart: always
+    image: wikijump/postgres
```
Comment (emmiegit) [inline install/docker-compose.aws-dev.yaml]: Important note: you're using `wikijump` as the org, but in our current docker docs we use `scpwiki`, and `scpwiki/postgres` is the one currently updated. I guess this would explain why the dev deploy currently live has an outdated seed database.

If we're going to move to `wikijump` we should aim to decommission the `scpwiki` one to prevent duplicative issues like this.

File: install/docker-compose.yaml [hunk 39]
```diff
@@ -27,9 +27,11 @@ services:
       - traefik.enable=true
       - traefik.http.routers.php-fpm.tls=true
       - traefik.http.routers.php-fpm.rule=Method(`GET`,`POST`,`HEAD`,`PUT`,`PATCH`,`DELETE`)  # Poor man's default route
+    # volumes:
+    #   - I:\Extra\wikijump:/var/www
```
Comment (emmiegit) [inline install/docker-compose.yaml]: Commented code.

File: install/php-fpm/wikijump.ini [hunk 40]
```diff
@@ -69,11 +69,11 @@ ssl = false
...
 ; TODO: have this be determined by the ssl var
 ; default: https
-schema = http
+schema = https
 
 ; [allow_http]
-; Purpose: Choose whether or not to allow ANY insecure traffic anywhere on your wiki farm. This is only recommended for
-;   local development where using SSL is tricky. The expectation in production is SSL is enabled everywhere.
+; Purpose: Choose whether or not to allow ANY insecure traffic anywhere on your wiki farm. You will need this set to true if
+; another service is handling SSL termination in front of Wikijump to avoid redirect loops.
```
Comment (emmiegit) [inline install/php-fpm/wikijump.ini:76]: :+1:

File: web/php/utils/GlobalProperties.php [hunk 41]
```diff
@@ -155,91 +155,91 @@ public static function init()
         self::$iniConfig = parse_ini_file(WIKIJUMP_ROOT . "/conf/wikijump.ini", true);
 
         // main settings
-        self::$WIKI_FARM                = self::fromIni("main", "wiki_farm", true);
+        self::$WIKI_FARM                = $_ENV["WIKI_FARM"] ?? self::fromIni("main", "wiki_farm", true);
```
Comment (emmiegit) [inline web/php/utils/GlobalProperties.php:158]: We should probably refactor this into a more generic configuration loader in the future, lol

File: web/web/heartbeat.php [hunk 42]
```diff
@@ -0,0 +1 @@
+<?php echo("ok")
```
Comment (emmiegit) [inline web/web/heartbeat.php:1]: lol

should change this to something more topical, like `;-)`

**Review by emmiegit — APPROVED** [review]

File: docs/deployment.md (hunk 1 above)
Comment (danieltharp) [reply docs/deployment.md]: An Elastic IP in AWS parlance is a public IP address that can be freely associated with a customer's assets, such as load balancers and EC2s. The IP is reserved by the account until explicitly released back to the pool so it functions as a static IP for a service. Without the usage of Elastic IP, every time the load balancer was recreated it would receive a new IP from the pool.

Comment (danieltharp) [reply docs/deployment.md]: The latter is the correct interpretation. I'll reword.

File: docs/deployment.md (hunk 2 above)
Comment (danieltharp) [reply docs/deployment.md]: There's an HTTP->HTTPS forced upgrade for requests explicitly made to HTTP assets at wjfiles.com.

File: docs/deployment.md (hunk 3 above)
Comment (danieltharp) [reply docs/deployment.md]: This is actually no longer relevant, no outside work is required for this deployment aside from storing secrets for the AWS creds to run this under. Updated in the next commit.

File: install/docker-compose.aws-dev.yaml (hunk 38 above)
Comment (danieltharp) [reply install/docker-compose.aws-dev.yaml]: This file is not needed at all, I will remove.

File: install/docker-compose.yaml (hunk 39 above)
Comment (danieltharp) [reply install/docker-compose.yaml]: Until we really firm up local dev environments I'll settle for changing this into an informational thing. Mapping a local volume for the web container is super duper useful.

File: web/php/utils/GlobalProperties.php (hunk 41 above)
Comment (danieltharp) [reply web/php/utils/GlobalProperties.php:158]: Yeah, agreed. The pattern will be essentially the same, I do want every facet here to be able to be overridden by environment variables but it could be presented in a nicer way.

File: web/web/heartbeat.php (hunk 42 above)
Comment (danieltharp) [reply web/web/heartbeat.php:1]: this requested change will not be implemented

File: infra/terraform/dev/deploy/.gitignore (hunk 4 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/.gitignore:15]: Yes, we are handling this a bit differently for now and people may want to manage their own terraform state their own way so I'd like to leave the guidance in.

File: infra/terraform/dev/deploy/.gitignore (hunk 5 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/.gitignore:29]: Same as above, we're not making use of this right now but the guidance for others is worth leaving this in since doing this part wrong will leak credentials.

File: infra/terraform/dev/deploy/cloudfront.tf (hunk 8 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/cloudfront.tf:78]: Code requests are going to the wikijump web containers and files aren't.

File: infra/terraform/dev/deploy/cloudwatch.tf (hunk 9 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/cloudwatch.tf:19]: Resolved in next commit.

File: infra/terraform/dev/deploy/ecs.tf (hunk 13 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/ecs.tf]: Resolved in next commit.

File: infra/terraform/dev/deploy/elb.tf (hunk 16 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/elb.tf]: Resolved in next commit, the line above was actually more pertinent.

File: infra/terraform/dev/deploy/iam.tf (hunk 17 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/iam.tf:11]: This is a known value from AWS, a globally available policy that has the exact permission set needed for this piece.

File: infra/terraform/dev/deploy/launch.tf (hunk 19 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/launch.tf]: This actually is bash with some interpolated HCL. It also seems to work as container metrics are flowing correctly in the cluster dashboard so I'd like to leave this alone until it's proven to be a bad idea.

File: infra/terraform/dev/deploy/cloudfront.tf (hunk 8 above)
Comment (emmiegit) [reply infra/terraform/dev/deploy/cloudfront.tf:78]: A potential issue here is that file requests are also gated by Wikidot's permissions model. If you don't have access to see a page, you also don't have access to see any files attached to that page.

File: infra/terraform/dev/deploy/launch.tf (hunk 19 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/launch.tf]: I'm going to try and remove the quotes entirely and see if it'll still interpolate correctly.

File: infra/terraform/dev/deploy/launch.tf (hunk 20 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/launch.tf:8]: Incorporated in next commit

File: infra/terraform/dev/deploy/providers.tf (hunk 21 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/providers.tf:8]: Resolved in next commit.

File: infra/terraform/dev/deploy/s3.tf (hunk 23 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/s3.tf:7]: Yeah, actual code and images and stuff in comparison to things like logs and other AWS byproducts.

File: infra/terraform/dev/deploy/task-definitions.tf (hunk 24 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/task-definitions.tf:43]: I need to work on this in a future PR once the GH Actions piece works. This piece should only be needed one time and then GH Actions will be dynamically updating the image to use with the (SHA) tag.

File: infra/terraform/dev/deploy/task-definitions.tf (hunk 25 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/task-definitions.tf]: Fair. Their example for this worked and I left it alone. Fixed in next commit.

File: infra/terraform/dev/deploy/wikijump.auto.tfvars (hunk 28 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/wikijump.auto.tfvars]: Included in next commit.

File: infra/terraform/dev/deploy/wikijump.auto.tfvars (hunk 29 above)
Comment (danieltharp) [reply infra/terraform/dev/deploy/wikijump.auto.tfvars:36]: `info@` is currently a valid mailbox and `security@` isn't, but something to do once there are more groups set up.

File: infra/terraform/prod/cloudfront.tf (hunk 32 above)
Comment (danieltharp) [reply infra/terraform/prod/cloudfront.tf:1]: I didn't mean to touch prod stuff in this, I'm not going to worry about it just at the moment since almost all of it's going to change eventually. Blame VS code for any files in prod that were touched.

File: infra/terraform/prod/vpc.tf (hunk 37 above)
Comment (danieltharp) [reply infra/terraform/prod/vpc.tf:49]: I didn't initially have any routing set up in here, what's available now is what's needed for the environment to run and at the moment I'm okay with dev running on a publicly available instance to have another way in aside from via the load balancer. If in the future the load balancer piece is so rock solid that I want to put the container EC2 in a private subnet the routes will change.
